### PR TITLE
Add present in window option for impress

### PR DESCRIPTION
**Pull request by NickWingate**
Add new 'presentinwindow' command which opens
a new window to display the slideshow svg in.

Change-Id: Ia28fcff0fd4ad64aca60381db202d7073661dc8a

* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Review by pedropintosilva — CHANGES_REQUESTED** [review]
Awesome!

# Plan

@NickWingate "I had a few ux questions"... 
@pedropintosilva : "let me answer inline ":

* Can the user open multiple presentation windows?
  * We shouldn't allow multiple presentations from the same document
  * User should be able to trigger other windows from other documents
* Can the user present in the original window while a new window is open?
  * We shouldn't allow multiple presentations from the same document even if one is fullscreen and the other is in a window
* Should escape close the new window (slightly harder to implement)
  * Even thought the user can simply press the native close button to close the windows, it would be nice to have the escape doing the same.
* If the user has blocked popups should we inform them or is the browser 'a popup has been blocked' enough?
  * I think it's not a must for now but I think we we should trigger 1 snackbar with 2 different messages depending on the context and [good to have) a full dialoginfo:
    *  [good to have] If the pop-up was blocked: Show dialoginfo with some sort of `Presentation was blocked. Please allow pop-ups in your browser. This lets slide shows to be displayed in separated windows, allowing for easy screen sharing.` + [image]
    * [we should] add ~~2~~ 1 snackbars
      * If the pop-up was not blocked: snackbar without timeout 
        * `Presenting slide show in window` [Dismiss] [Close Presentation]
# Errors on FF

<details>
I'm getting the following whenever I press the regular fullscreen button:
```javascript
14:11:30.197 1697803890197 INCOMING: jsdialog: {"id":"busypopup","jsontype":"dialog","type":"modalpopup","action":"fadeout"} cool.html:357:37
14:11:30.197 1697803890198 OUTGOING: downloadas name=slideshow.svg id=slideshow format=svg options= cool.html:357:37
14:11:30.205 InstallTrigger is deprecated and will be removed in the future. constants.js:50:14
14:11:30.272 1697803890272 OUTGOING: clientvisiblearea x=0 y=-960 width=9795 height=13830 splitx=0 splity=0 cool.html:357:37
14:11:30.280 1697803890281 INCOMING: downloadas: downloadid=ZQiaknqK7sAy4OsJFVlWLI3EjQQD0p72RsiNLc28SBXSeqe8SIfuObsvbsgn5FNW port=9980 id=slideshow cool.html:357:37
14:11:30.281 1697803890281 INCOMING: jsdialog: {"id":"busypopup","jsontype":"dialog","type":"modalpopup","action":"fadeout"} cool.html:357:37
14:11:30.281 slide file url :  http://localhost:9980/cool/http%3A%2F%2Flocalhost%3A9980%2Fwopi%2Ffiles%2Fhome%2Fpedro%2FLibreOffice%2Fonline-master%2Ftest%2Fdata%2Fhello-world.odp/download/ZQiaknqK7sAy4OsJFVlWLI3EjQQD0p72RsiNLc28SBXSeqe8SIfuObsvbsgn5FNW?WOPISrc=http://localhost:9980/wopi/files/home%2Fpedro%2FLibreOffice%2Fonline-master%2Ftest%2Fdata%2Fhello-world.odp&compat=/ws cool.html:357:37
14:11:30.337 InstallTrigger is deprecated and will be removed in the future. constants.js:50:14
14:11:30.388 Error: we do not count styles for frames msg.js:84:19
14:11:30.395 XML Parsing Error: not well-formed
Location: 
Line Number 17, Column 51:
14:11:30.398 SyntaxError: An invalid or illegal string was specified content-script.js:39
```

and sometimes just
```javascript
14:17:10.968 Error: we do not count styles for frames msg.js:84:19

```

and sometimes, but probably an old bug:
```javascript
14:19:57.888 Uncaught (in promise) TypeError: Fullscreen request denied
    _onFullScreen Map.SlideShow.js:53
    fire Events.js:152
    dispatch Toolbar.js:1095
    _mapBigDispatchToolItem Control.JSDialogBuilder.js:2669
    jQuery 9
    _mapBigDispatchToolItem Control.JSDialogBuilder.js:2668
    build Control.NotebookbarBuilder.js:1184
    build Control.NotebookbarBuilder.js:1193
    _tabPageHandler Control.JSDialogBuilder.js:651
    _overriddenTabPageHandler Control.NotebookbarBuilder.js:356
    build Control.NotebookbarBuilder.js:1184
    _tabsControlHandler Control.JSDialogBuilder.js:1379
    _overriddenTabsControlHandler Control.NotebookbarBuilder.js:352
    build Control.NotebookbarBuilder.js:1184
    build Control.NotebookbarBuilder.js:1193
    build Control.NotebookbarBuilder.js:1193
    loadTab Control.Notebookbar.js:222
    onAdd Control.Notebookbar.js:45
    addControl Control.js:98
    createNotebookbarControl Control.UIManager.js:441
    refreshNotebookbar Control.UIManager.js:460
    refreshUI Control.Zotero.js:140
    updateUserID Control.Zotero.js:157
    promise callback*updateUserID Control.Zotero.js:154
    onUpdateViews Control.UIManager.js:866
    fire Events.js:152
    _onViewInfoMsg CanvasTileLayer.js:2969
    _onMessage CanvasTileLayer.js:1763
    _onMessage Socket.js:1197
    _emitSlurpedEvents Socket.js:347
    _slurpTimer Socket.js:297
    setTimeout handler*_queueSlurpEventEmission Socket.js:295
    _slurpMessage Socket.js:395
    connect Socket.js:64
    loadDocument Map.js:321
    <anonymous> main.js:93
    <anonymous> main.js:113
Map.SlideShow.js:53:20

```

</details>

**Comment by NickWingate** [comment]
## Progress
#### Features
- I've prevented multiple presentations from the same document - just with an alert for now but will change to a vex dialog or snackbar
- Implemented `Esc` closes windowed presentation
#### Bugs
- Fixed one of `Fullscreen Denied` FF bugs when the parent is unfocused
- Found another `Fullscreen Denied` bug due to fullscreen not being called fast enough after user event
	- `Request for fullscreen was denied because Element.requestFullscreen() was not called from inside a short running user-generated event handler. Util.js:182:29`
	- Think time limit is 1 second from this SO post: https://stackoverflow.com/a/23913497
- Haven't been able to reproduce other two bugs: `InstallTrigger...` and `14:17:10.968 Error: we do not count styles for frames msg.js:84:19`

## Todo
- [ ] Check for mobile and dont allow
- [ ] Popup blocked snackbars
- [ ] Convert alerts to vex dialogs

**Comment by pedropintosilva** [comment]
@NickWingate thanks. Note as described in https://github.com/CollaboraOnline/online/pull/7493#pullrequestreview-1689989201 the dialog case is a low priority. I have also reword it to make it more clear (TLDR: we only need one snackbar). For the `dialoginfo` I mention above, I'm referring to a specific JSDialog type ( we don't use vex anymore)

File: browser/src/control/Control.UIManager.js
Comment (eszkadev) [inline browser/src/control/Control.UIManager.js:945]: we can reuse "action" for that, no need for new parameter

**Review by pedropintosilva — APPROVED** [review]
Tested, looks good, but I would like to have the +1 from @eszkadev 

I also see that css change to target every id that ends with `button` I wonder why it was necessary but then again maybe makes more sense than what we had before

**Comment by pedropintosilva** [comment]
if @eszkadev approves we can merge, I'm now preparing the missing icon in another additional PR (so to avoid re-running CI)

**Comment by pedropintosilva** [comment]
https://github.com/CollaboraOnline/online/pull/8003

**Comment by NickWingate** [comment]
> I also see that css change to target every id that ends with `button` I wonder why it was necessary but then again maybe makes more sense than what we had before

Changed it to allow snackbar to have two buttons with different ids but the same style:
`withDismiss ? {id: 'snackbar-dismiss-button', type: 'pushbutton', text: _('Dismiss')} : {},`  --  `Control.UIManager.js:1053`

so that we can differentiate which button was pressed on the snackbar:
`if (object.id === 'snackbar-dismiss-button' && objectType === 'pushbutton' && eventType === 'click')` -- `Control.UIManager.js:1074`

File: browser/src/map/handler/Map.SlideShow.js
Comment (eszkadev) [inline browser/src/map/handler/Map.SlideShow.js:172]: last param is not used, showSnackbar has only 6 params

**Review by eszkadev — APPROVED** [review]
From code point of view it looks ok, we can merge, I will prepare followup with removal of unused param. @pedropintosilva feel free to merge if you think it is important

**Comment by eszkadev** [comment]
followup: https://github.com/CollaboraOnline/online/pull/8010

**Comment by pedropintosilva** [comment]
Thanks @eszkadev for the follow up. But didn't you meant to push the additional commit here? Or you preferred like that in the case @NickWingate has some other comment and this PR ends up to be still useful?

**Review by pedropintosilva — CHANGES_REQUESTED** [review]
do not merge this, follow up PR with the additional last commit (that removes the extra snackbar parameter) : https://github.com/CollaboraOnline/online/pull/8010

**Comment by eszkadev** [comment]
followup merged

**Pull request closed without merging**
